File: 141-linked-list-cycle/linked-list-cycle.swift
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public var val: Int
 *     public var next: ListNode?
 *     public init(_ val: Int) {
 *         self.val = val
 *         self.next = nil
 *     }
 * }е
 */

class Solution {

    func hasCycle(_ head: ListNode?) -> Bool {
        
        var slow: ListNode? = head
        var fast: ListNode? = head?.next

        while fast != nil {

            if slow === fast {
                return true
            }

            slow = slow?.next
            fast = fast?.next?.next
        }

        return false
    }
}

File: 141-linked-list-cycle/linked-list-cycle.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    
    private final static ListNode next(ListNode node) {
        if (node == null) {
            return null;
        }
        
        return node.next;
    }
    
    public boolean hasCycle(ListNode head) {
        
        var slowPointer = head;
        var fastPointer = head;
        
        while (true) {
            
            slowPointer = next(slowPointer);
            fastPointer = next(next(fastPointer));
            
            if (fastPointer == null) {
                return false;
            }
            
            if (slowPointer == fastPointer) {
                return true;
            }
        }
    }
}